File: param_tutorial/src/calculate_server.cpp
#include "ros/ros.h"
#include "param_tutorial/Calculate.h" // 서비스 헤더 파일. 빌드시 생성

#define PLUS 1
#define MINUS 2
#define MULTIPLICATION 3
#define DIVISION 4

int my_operator = PLUS;

// 콜백함수 
bool calculation(param_tutorial::Calculate::Request& req, param_tutorial::Calculate::Response& res){
    switch(my_operator)
    {
        case PLUS:
            res.result = req.a + req.b;
            break;
        case MINUS:
            res.result = req.a - req.b;
            break;
        case MULTIPLICATION:
            res.result = req.a * req.b;
            break;
        case DIVISION:
            res.result = req.a / req.b;
            break;
        default: 
            res.result = req.a + req.b;
            break;
    }

    ROS_INFO("a: %d, b: %d, result: %d", req.a, req.b, res.result);
    
    return true;
}

int main(int argc, char** argv){

    ros::init(argc, argv, "calculate_server"); // 노드 이름 초기화
    ros::NodeHandle nh;

    // 파라미터 서버에 calculation_method라는 변수의 이름으로 PLUS 값을 넣는다
    nh.setParam("calculation_method", PLUS); // 파라미터 초기 설정

    // 서비스 서버 (calculate_server) 선언
    // 서비스 이름은 (calculate)이고, 요청이 왔을 때 콜백함수 (calculation)를 실행한다.
    ros::ServiceServer calculate_server = nh.advertiseService("calculate", calculation);

    ros::Rate loop_rate(10); // 0.1초마다

    while(ros::ok()){ // ros 켜져있는 동한 무한 반복
        // 파라미터 서버에 있는 calculation_method라는 이름의 변수를 my_operator에 가져와서 할당(저장)
        nh.getParam("calculation_method", my_operator);
        // 대기하다가 콜백함수 한번만 호출. spin 한번만하고 꺼짐
        ros::spinOnce();
        loop_rate.sleep(); // 주기에 맞게 sleep
    }

    return 0;
}

File: param_tutorial/scripts/calculate_server.py
#!/usr/bin/python3
#-*- coding:utf-8 -*-
from asyncore import read
import rospy
from param_tutorial.srv import Calculate, CalculateResponse

PLUS = 1
MINUS = 2
MULTIPLICATION = 3
DIVISION = 4

# 콜백함수
def calculation(req):
    my_operator = rospy.get_param("calculation_method")
    if my_operator == PLUS:
        result = req.a + req.b
    elif my_operator == MINUS:
        result = req.a - req.b
    elif my_operator == MULTIPLICATION:
        result = req.a * req.b
    elif my_operator == DIVISION:
        # if req.b == 0:
        #     result = 0
        # else: 
        #     result  = req.a // req.b
        try:
            result = int(req.a / req.b)
        except ZeroDivisionError:
            result = 0
    else: 
        result = req.a + req.b

    rospy.loginfo(f"a: {req.a}, b: {req.b}, reslut: {result}")
    
    res = CalculateResponse()
    res.result = result

    return res

def calculate_server():
    rospy.init_node("calculate_server") # 노드 이름 초기화

    # 마스터에 있는 파라미터 서버에 있는 변수 calculation_method를 PLUS로 set
    rospy.set_param("calculation_method", PLUS) 
    
    # 서비스 서버 (my_server)선언
    # 서비스 이름(calculate), 서비스 타입(Calculate), 콜백함수(calculation)
    my_server = rospy.Service("calculate", Calculate, calculation)
    rospy.loginfo("Service Server ready")

    rospy.spin()


if __name__ == "__main__":
    calculate_server()